File: combinationsum.go
// Time Complexity : O(n raised to m choices)
// Space Complexity : O(n)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this :

// Your code here along with comments explaining your approach
/*
in recursive function set base case if sum is target append to res if its greater return
for each candidate
append current to temp
call recursive with new sum, new temp and new index
remove current from temp
*/
package main

import "fmt"

var temp []int
var res [][]int
func combinationSum(candidates []int, target int) [][]int {
	temp = []int{}
	res = [][]int{}
	h(candidates, target, 0, 0, temp)
	return res
}

func h(candidates []int, target int, index int, sum int, temp []int) {
	//base
	if target == sum {
		temp1 := []int{}
		for x:=0;x <len(temp);x++{
			temp1 = append(temp1, temp[x])
		}
		res = append(res, temp1)
		return
	}
	if sum > target {
		return
	}

	for i:=index;i<len(candidates);i++ {
		temp = append(temp, candidates[i])
		h(candidates, target, i, sum + candidates[i], temp)
		temp = temp[:len(temp)-1]
	}
}

func MainCombinationSum() {
	fmt.Println(combinationSum([]int{2,3,6,7}, 7)) //expected [[2,2,3],[7]]
}

File: combinationSum.swift
// Time Complexity : 
// Space Complexity : 
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
    var output = [[Int]]()
    func combinationSum(_ candidates: [Int], _ target: Int) -> [[Int]] {
        if candidates == nil || candidates.count == 0 {
            return []
        }
        var path = [Int]()
        backtracking(candidates,target,0,&path)
        return output
        
    }
    func backtracking(_ candidates: [Int],_ target: Int,_ index: Int,_  path: inout [Int] ) {
        if target < 0 {
            return
        }
        if target == 0 {
            var path = path
            output.append(path)
        }
        for i in index..<candidates.count {
            path.append(candidates[i])
            backtracking(candidates, target - candidates[i],i,&path)
            path.remove(at: path.count-1)
        }
    }